src/core/packages/http/router-server-internal/src/route.ts

```diff
@@ -21,3 +53,162 @@ export type InternalRouteConfig<P, Q, B, M extends RouteMethod> = Omit<
+    kibanaRequest.apiVersion = version;
+  } catch (error) {
+    kibanaRequest = CoreKibanaRequest.from(request);
+    kibanaRequest.apiVersion = version;
+
+    log.error('400 Bad Request', formatErrorMeta(400, { request, error }));
+
+    const response = kibanaResponseFactory.badRequest({
+      body: error.message,
+      headers: isPublicAccessApiRoute(routeInfo)
+        ? getVersionHeader(BASE_PUBLIC_VERSION)
+        : undefined,
+    });
+    return { error: response };
+  } finally {
+    router.emitPostValidate(
+      kibanaRequest!,
+      getPostValidateEventMetadata(kibanaRequest!, routeInfo)
+    );
+  }
+
+  return { ok: kibanaRequest };
+}
+
+/** @internal */
+export const handle = async (
+  request: Request,
+  { router, route, handler, routeSchemas, log }: HandlerDependencies
+) => {
+  const { error, ok: kibanaRequest } = validateHapiRequest(request, {
+    routeInfo: {
+      access: route.options?.access,
+      httpResource: route.options?.httpResource,
+      deprecated: route.options?.deprecated,
+    },
+    router,
+    log,
+    routeSchemas,
+  });
+  if (error) {
+    return error;
+  }
+  const kibanaResponse = await handler(kibanaRequest, kibanaResponseFactory);
+  if (isPublicAccessApiRoute(route.options)) {
+    injectVersionHeader(BASE_PUBLIC_VERSION, kibanaResponse);
+  }
+  return kibanaResponse;
+};
+
+function isPublicAccessApiRoute({
+  access,
+  httpResource,
+}: {
+  access?: RouteAccess;
+  httpResource?: boolean;
+} = {}): boolean {
+  return !httpResource && access === 'public';
+}
+
+/**
+ * Create the validation schemas for a route
+ *
+ * @returns Route schemas if `validate` is specified on the route, otherwise
+ * undefined.
+ */
+function routeSchemasFromRouteConfig<P, Q, B>(
+  route: InternalRouteConfig<P, Q, B, typeof routeMethod>,
+  routeMethod: RouteMethod
+) {
+  // The type doesn't allow `validate` to be undefined, but it can still
+  // happen when it's used from JavaScript.
+  if (route.validate === undefined) {
+    throw new Error(
+      `The [${routeMethod}] at [${route.path}] does not have a 'validate' specified. Use 'false' as the value if you want to bypass validation.`
+    );
+  }
+
+  if (route.validate !== false) {
+    const validation = getRequestValidation(route.validate);
+    Object.entries(validation).forEach(([key, schema]) => {
+      if (!(isConfigSchema(schema) || isZod(schema) || typeof schema === 'function')) {
+        throw new Error(
+          `Expected a valid validation logic declared with '@kbn/config-schema' package, '@kbn/zod' package or a RouteValidationFunction at key: [${key}].`
+        );
+      }
+    });
+    return RouteValidator.from(validation);
+  }
+}
+
+function getPostValidateEventMetadata(request: AnyKibanaRequest, routeInfo: RouteInfo) {
+  return {
+    deprecated: routeInfo.deprecated,
+    isInternalApiRequest: request.isInternalApiRequest,
+    isPublicAccess: routeInfo.access === 'public',
```

Comment:
nit: Do we want to include this change in the backport? Or do we prefer to backport it as-is and later backport https://github.com/elastic/kibana/pull/207829?

Reply:
Would be nice if we can include it in one swoop :)